torch_xla/_internal/tpu.py

```diff
@@ -268,23 +270,50 @@ def configure_topology(local_rank: int,
 
 
 def discover_master_worker_ip(use_localhost: bool = True) -> str:
-  """Find the IP of the TPU host with TPU:0.
+  """Find the IP of the master TPU host.
+
+  In multiprocess, this is the host with TPU:0.
+  In SPMD mode, this is the host running process 0.
 
   TPU device IDs are nondeterministic and independent from Cloud TPU worker IDs.
 
   Args:
     use_localhost: if there is only one TPU host, return 'localhost` instead
       of that host's internal IP.
   """
+  import torch_xla.runtime as xr
   worker_ips = get_worker_ips()
   if len(worker_ips) == 1:
     return 'localhost'
 
   tpu_env = get_tpu_env()
   current_worker_id = int(tpu_env[xenv.WORKER_ID])
+  if xr.is_spmd():
+    return _spmd_find_master_ip(worker_ips[current_worker_id])
+
   t = torch.tensor([current_worker_id], device=xm.xla_device())
   xm.collective_broadcast([t])
   xm.mark_step()
 
   master_worker_id = int(t.cpu())
   return worker_ips[master_worker_id]
+
+
+def _spmd_find_master_ip(current_worker_ip: str) -> str:
+  import torch_xla.runtime as xr
+  import torch_xla.experimental.xla_sharding as xs
+  from_cpu_shards = torch_xla._XLAC._global_tensor_from_cpu_shards
+  ip_int = int(ip_address(current_worker_ip))
+  n_dev = xr.global_runtime_device_count()
+  local_ndev = len(torch_xla._XLAC._xla_get_runtime_devices())
+  # Create a global (n_dev x 2) tensor containing all process indices and IPs,
+  # and find the process 0 IP as the master IP.
+  shard = torch.LongTensor([[xr.process_index(), ip_int]])
```

Comment:
Could we use global device ordinal instead of the process idx? It's the same thing, but wonder if it's feasible -- that way your search for the IP associated with TPU0 would be more consistent with the code?

Reply:
Unfortunately it's actually different - there's no guaranteed mapping between processes and devices, so process 0 may not control device 0 and we can't pull the IP out of the device 0 slot in the global tensor. In SPMD, I think it's better to rely on the process abstraction, since we never actually need the devices for collectives.

Reply:
Thanks @jonb377 , we are not looking for the IP associated with TPU0, but with the rank=0.